File: android/app/src/test/java/de/youspeed/android/alpha/LaneDetectionTests.kt
package de.youspeed.android.alpha

import kotlin.math.abs
import org.junit.Assert.assertEquals
import org.junit.Assert.assertNotNull
import org.junit.Assert.assertNull
import org.junit.Assert.assertTrue
import org.junit.Test

class LaneDetectionTests {
    private val detector = LaneDetector()

    private fun road(left: Boolean = true, right: Boolean = true, width: Int = 384, height: Int = 216): ByteArray {
        val image = ByteArray(width * height) { 55 }
        for (y in 0 until height) {
            val normalizedY = y.toDouble() / (height - 1)
            if (normalizedY !in 0.52..0.96) continue
            val offset = (normalizedY - 0.52) * 0.30 / 0.44
            for (x in 0 until width) {
                val normalizedX = x.toDouble() / (width - 1)
                if ((left && abs(normalizedX - (0.42 - offset)) < 0.0065) ||
                    (right && abs(normalizedX - (0.58 + offset)) < 0.0065)) {
                    image[y * width + x] = 230.toByte()
                }
            }
        }
        return image
    }

    @Test
    fun pairNeedsTwoFramesAndNeverInventsMissingBoundary() {
        val tracker = LaneTracker()
        val first = detector.detect(road(), 384, 216, 1.0)
        assertTrue(first.hasReliablePair)
        assertEquals(LaneDetectionState.UNCERTAIN, tracker.update(first).state)
        assertTrue(tracker.update(detector.detect(road(), 384, 216, 1.2)).hasReliablePair)
        val single = tracker.update(detector.detect(road(right = false), 384, 216, 1.4))
        assertEquals(LaneDetectionState.UNCERTAIN, single.state)
        assertNotNull(single.left)
        assertNull(single.right)
        val blank = tracker.update(detector.detect(ByteArray(384 * 216) { 55 }, 384, 216, 1.6))
        assertEquals(LaneDetectionState.UNAVAILABLE, blank.state)
        assertNull(blank.left)
        assertNull(blank.right)
    }

    @Test
    fun backwardClockGapsAndExplicitResetRequireFreshConfirmation() {
        val tracker = LaneTracker()
        fun estimate(time: Double) = detector.detect(road(), 384, 216, time)
        tracker.update(estimate(1.0))
        assertTrue(tracker.update(estimate(1.2)).hasReliablePair)
        assertEquals(LaneDetectionState.UNCERTAIN, tracker.update(estimate(2.0)).state)
        assertTrue(tracker.update(estimate(2.2)).hasReliablePair)
        assertEquals(LaneDetectionState.UNCERTAIN, tracker.update(estimate(2.1)).state)
        tracker.reset()
        assertEquals(LaneDetectionState.UNCERTAIN, tracker.update(estimate(2.3)).state)
    }

    @Test
    fun invalidInputAndNonFiniteTimestampAreUnavailable() {
        listOf(Triple(0, 216, ByteArray(0)), Triple(641, 216, ByteArray(0)), Triple(384, 216, byteArrayOf(55))).forEach { (width, height, pixels) ->
            assertEquals(LaneDetectionState.UNAVAILABLE, detector.detect(pixels, width, height, 1.0).state)
        }
        assertEquals(LaneDetectionState.UNAVAILABLE, detector.detect(road(), 384, 216, Double.NaN).state)
        val invalid = LaneDetectionEstimate(null, null, Double.POSITIVE_INFINITY, LaneDetectionState.RELIABLE)
        assertEquals(LaneDetectionState.UNAVAILABLE, LaneTracker().update(invalid).state)
    }

    @Test
    fun fitStaysOnObservedMarkings() {
        val estimate = detector.detect(road(), 384, 216, 1.0)
        listOf(requireNotNull(estimate.left) to true, requireNotNull(estimate.right) to false).forEach { (boundary, isLeft) ->
            assertTrue(boundary.confidence >= 0.62)
            boundary.points.forEach { point ->
                assertTrue(point.y in 0.52..0.96)
                val offset = (point.y - 0.52) * 0.30 / 0.44
                assertEquals(if (isLeft) 0.42 - offset else 0.58 + offset, point.x, 0.008)
            }
        }
    }

    @Test
    fun corridorFillUsesOnlySharedObservedExtent() {
        val left = LaneBoundary(listOf(LanePoint(0.4, 0.5), LanePoint(0.1, 0.9)), 0.9)
        val right = LaneBoundary(listOf(LanePoint(0.6, 0.6), LanePoint(0.8, 0.8)), 0.9)
        val estimate = LaneDetectionEstimate(left, right, 1.0, LaneDetectionState.RELIABLE)
        assertEquals(4, estimate.corridorPoints.size)
        assertTrue(estimate.corridorPoints.all { it.y in 0.6..0.8 })
        assertEquals(0.325, estimate.corridorPoints.first().x, 1e-9)
        assertEquals(0.175, estimate.corridorPoints[1].x, 1e-9)
        assertTrue(estimate.copy(state = LaneDetectionState.UNCERTAIN).corridorPoints.isEmpty())
        assertTrue(estimate.copy(right = null).corridorPoints.isEmpty())
    }
}

File: iphone/SpeedConsumerApp/LaneDetectionTests.swift
import XCTest
@testable import SpeedConsumer

final class LaneDetectionTests: XCTestCase {
    private let detector = LaneDetector()

    private func road(left: Bool = true, right: Bool = true, width: Int = 384, height: Int = 216) -> [UInt8] {
        var image = [UInt8](repeating: 55, count: width * height)
        for y in 0..<height {
            let normalizedY = Double(y) / Double(height - 1)
            guard (0.52...0.96).contains(normalizedY) else { continue }
            let offset = (normalizedY - 0.52) * 0.30 / 0.44
            for x in 0..<width {
                let normalizedX = Double(x) / Double(width - 1)
                if (left && abs(normalizedX - (0.42 - offset)) < 0.0065)
                    || (right && abs(normalizedX - (0.58 + offset)) < 0.0065) {
                    image[y * width + x] = 230
                }
            }
        }
        return image
    }

    func testPairNeedsTwoFramesAndNeverInventsMissingBoundary() {
        var tracker = LaneTracker()
        let first = detector.detect(grayscale: road(), width: 384, height: 216, timestampSeconds: 1)
        XCTAssertTrue(first.hasReliablePair)
        XCTAssertEqual(tracker.update(first).state, .uncertain)
        XCTAssertTrue(tracker.update(detector.detect(grayscale: road(), width: 384, height: 216, timestampSeconds: 1.2)).hasReliablePair)
        let single = tracker.update(detector.detect(grayscale: road(right: false), width: 384, height: 216, timestampSeconds: 1.4))
        XCTAssertEqual(single.state, .uncertain)
        XCTAssertNotNil(single.left)
        XCTAssertNil(single.right)
        let blank = tracker.update(detector.detect(grayscale: [UInt8](repeating: 55, count: 384 * 216), width: 384, height: 216, timestampSeconds: 1.6))
        XCTAssertEqual(blank.state, .unavailable)
        XCTAssertNil(blank.left)
        XCTAssertNil(blank.right)
    }

    func testBackwardClockGapsAndExplicitResetRequireFreshConfirmation() {
        var tracker = LaneTracker()
        func estimate(_ timestamp: Double) -> LaneDetectionEstimate {
            detector.detect(grayscale: road(), width: 384, height: 216, timestampSeconds: timestamp)
        }
        _ = tracker.update(estimate(1))
        XCTAssertTrue(tracker.update(estimate(1.2)).hasReliablePair)
        XCTAssertEqual(tracker.update(estimate(2.0)).state, .uncertain)
        XCTAssertTrue(tracker.update(estimate(2.2)).hasReliablePair)
        XCTAssertEqual(tracker.update(estimate(2.1)).state, .uncertain)
        tracker.reset()
        XCTAssertEqual(tracker.update(estimate(2.3)).state, .uncertain)
    }

    func testInvalidInputAndNonFiniteTimestampAreUnavailable() {
        for (width, height, pixels) in [(0, 216, [UInt8]()), (641, 216, [UInt8]()), (384, 216, [55])] {
            XCTAssertEqual(detector.detect(grayscale: pixels, width: width, height: height, timestampSeconds: 1).state, .unavailable)
        }
        XCTAssertEqual(detector.detect(grayscale: road(), width: 384, height: 216, timestampSeconds: .nan).state, .unavailable)
        var tracker = LaneTracker()
        let invalid = LaneDetectionEstimate(left: nil, right: nil, timestampSeconds: .infinity, state: .reliable)
        XCTAssertEqual(tracker.update(invalid).state, .unavailable)
    }

    func testFitStaysOnObservedMarkings() throws {
        let estimate = detector.detect(grayscale: road(), width: 384, height: 216, timestampSeconds: 1)
        for (boundary, isLeft) in [(try XCTUnwrap(estimate.left), true), (try XCTUnwrap(estimate.right), false)] {
            XCTAssertGreaterThanOrEqual(boundary.confidence, 0.62)
            for point in boundary.points {
                XCTAssertTrue((0.52...0.96).contains(point.y))
                let offset = (point.y - 0.52) * 0.30 / 0.44
                XCTAssertEqual(point.x, isLeft ? 0.42 - offset : 0.58 + offset, accuracy: 0.008)
            }
        }
    }

    func testCorridorFillUsesOnlySharedObservedExtent() {
        let left = LaneBoundary(points: [LanePoint(x: 0.4, y: 0.5), LanePoint(x: 0.1, y: 0.9)], confidence: 0.9)
        let right = LaneBoundary(points: [LanePoint(x: 0.6, y: 0.6), LanePoint(x: 0.8, y: 0.8)], confidence: 0.9)
        let estimate = LaneDetectionEstimate(left: left, right: right, timestampSeconds: 1, state: .reliable)
        XCTAssertEqual(estimate.corridorPoints.count, 4)
        XCTAssertTrue(estimate.corridorPoints.allSatisfy { (0.6...0.8).contains($0.y) })
        XCTAssertEqual(estimate.corridorPoints[0].x, 0.325, accuracy: 1e-9)
        XCTAssertEqual(estimate.corridorPoints[1].x, 0.175, accuracy: 1e-9)
        XCTAssertTrue(LaneDetectionEstimate(left: left, right: right, timestampSeconds: 1, state: .uncertain).corridorPoints.isEmpty)
        XCTAssertTrue(LaneDetectionEstimate(left: left, right: nil, timestampSeconds: 1, state: .reliable).corridorPoints.isEmpty)
    }
}
